### PR TITLE
RH, KLDiv, Vanilla RHF, Sandu's experiments

### PR DESCRIPTION
**Pull request by Abhat-93**
Initial stuff. Not ready to merge

### CODE REVIEW

File: src/+datools/+examples/+sandu/l63_experiments.m [hunk 1]
```diff
@@ -13,25 +14,25 @@
...
 % create an array of rejuvination
-rejs = 2*logspace(-2, -1, 4);
-rejs = round(rejs,2);
+rejs = 2 * logspace(-2, -1, 4);
+rejs = round(rejs, 2);
 
 % define steps and spinups
-spinup = 500;
-times = 11*spinup;
+spinup = 50;
```
Comment (AndreyAPopov) [inline src/+datools/+examples/+sandu/l63_experiments.m:27]: why did you change this to 50? at least 100 IMO

File: src/+datools/+examples/+sandu/l63_experiments.m [hunk 2]
```diff
@@ -44,16 +45,16 @@
     case 'RHF'
         filtertype = 'Ensemble';
 end
-fprintf('Filtername = %s, Observation Variance = %.2f, Runs = %d, spinups = %d\n',...
+fprintf('Filtername = %s, Observation Variance = %.2f, Runs = %d, spinups = %d\n', ...
     filtername, variance, times, spinup);
 % time steps
-Delta_t = 0.12; 
+Delta_t = 0.12;
```
Comment (AndreyAPopov) [inline src/+datools/+examples/+sandu/l63_experiments.m:51]: Deltat

Comment (AndreyAPopov) [reply src/+datools/+examples/+sandu/l63_experiments.m:51]: or Dt

File: src/+datools/+examples/+sandu/l63_experiments.m [hunk 3]
```diff
@@ -62,24 +63,23 @@
...
     speye(natureODE.NumVars));
 
 % observe these variables
 % change the array if you want to observe lesser state variables
 observeindicies = 1:1:natureODE.NumVars;
-
 nobsvars = numel(observeindicies);
 
-R = variance*(1/1)*speye(nobsvars);
+R = variance * (1 / 1) * speye(nobsvars);
```
Comment (AndreyAPopov) [inline src/+datools/+examples/+sandu/l63_experiments.m:82]: why is there a 1/1???

File: src/+datools/+examples/+sandu/l63_experiments.m [hunk 4]
```diff
@@ -189,213 +191,234 @@
...
                 rmse = sqrt(mean(mses(1:(i - spinup))));
-                
+
                 rmstempval(i - spinup) = rmse;
-                
-%                 if rmse > maxallowerr || isnan(rmse) || mses(i - spinup) > 2*maxallowerr
-%                     do_enkf = false;
-%                 end
+
+                %                 if rmse > maxallowerr || isnan(rmse) || mses(i - spinup) > 2*maxallowerr
```
Comment (AndreyAPopov) [inline src/+datools/+examples/+sandu/l63_experiments.m:246]: why was this commented out?

File: src/+datools/+examples/+sandu/l63_experiments.m [hunk 5]
```diff
@@ -189,213 +191,234 @@
...
+                %                     do_enkf = false;
+                %                 end
             end
-            
+
             if ~do_filter
```
Comment (AndreyAPopov) [inline src/+datools/+examples/+sandu/l63_experiments.m:251]: I know I did do_enkf, but please make this dofilter or bdofilter

File: src/+datools/+examples/+sandu/l96_experiments.m [hunk 6]
```diff
@@ -54,16 +54,16 @@
     case 'RHF'
         filtertype = 'Ensemble';
 end
-fprintf('Filtername = %s, Observation Variance = %.2f, Runs = %d, spinups = %d\n',...
+fprintf('Filtername = %s, Observation Variance = %.2f, Runs = %d, spinups = %d\n', ...
     filtername, variance, times, spinup);
 % time steps
-Delta_t = 0.05; 
+Delta_t = 0.05;
```
Comment (AndreyAPopov) [inline src/+datools/+examples/+sandu/l96_experiments.m:60]: Deltat


Just search for "_" and do all those vars correctly

File: src/+datools/+statistical/+ensemble/Hybrid.m [hunk 7]
```diff
@@ -37,7 +37,7 @@ function analysis(obj, R, y)
             X = obj.Ensemble;
             alphas = obj.Alphas;
             
-            Xf = X;
+            xf = X;
```
Comment (AndreyAPopov) [inline src/+datools/+statistical/+ensemble/Hybrid.m:40]: I don't understand why this was made lowercase. I know matlab variables are lowercase, but in our notation (which ALWAYS overrides matlab notation) Xf stands for an ensemble, and xf stands for a particular realization (so a vector not a matrix). Please change this back.

File: src/+datools/+statistical/+ensemble/RHF.m [hunk 8]
```diff
@@ -1,158 +1,167 @@
...
             tc = obj.Model.TimeSpan(1);
-            
-            % define length of gaussian tail
-            gtl = obj.Truncate;
-            discrete_pts = obj.DiscretePoint;
-            
+
+            % define length of gaussian tail (not needed now)
+            %gtl = obj.Truncate;
+            %discrete_pts = obj.DiscretePoint;
```
Comment (AndreyAPopov) [inline src/+datools/+statistical/+ensemble/RHF.m:35]: If it's not needed, why is it still here. delete

File: src/+datools/+statistical/+ensemble/RHF.m [hunk 9]
```diff
@@ -1,158 +1,167 @@
...
-                    muright = xfs(end) - sqrt(2*Var)*erfinv(1 - 2/(ensN + 1));
-                    mu = [muleft muright];
+                    % find the mean of the left gaussian tail
+                    muleft = xfs(1) - sqrt(2*Var) * erfinv(2/(ensN + 1)-1);
+                    % find the mean of the right gaussian tail
+                    muright = xfs(end) - sqrt(2*Var) * erfinv(1-2/(ensN + 1));
+                    mu = [muleft, muright];
                 else
-                    fprintf('Error! Gaussian tail not provided!\n');
+                    fprintf('Error! Only Gaussian Tail is supported as of now!\n');
```
Comment (AndreyAPopov) [inline src/+datools/+statistical/+ensemble/RHF.m:82]: if this is an error, call an error...

`error('Only Gaussian Tail is supported as of now.');`

File: src/+datools/+statistical/+ensemble/RHF.m [hunk 10]
```diff
@@ -1,158 +1,167 @@
...
                     return;
                 end
-                                
-                inn = y(i) - sort(Hxf(i,:));
+
+                % calculate the likelihood for this state
+                innovation = y(i) - sort(Hxf(i, :));
```
Comment (AndreyAPopov) [inline src/+datools/+statistical/+ensemble/RHF.m:87]: we usually use d for the innovation.

File: src/+datools/+statistical/+ensemble/RHF.m [hunk 11]
```diff
@@ -1,158 +1,167 @@
...
                 ntemp = length(y(i));
-                likelihood = (diag((1/sqrt((2*pi)^ntemp*det(R(i,i)))) * ...
-                    exp(-0.5*inn.'*(R(i,i)\inn)))).';
-                likelihood = likelihood/min(likelihood); % for scaling
-                l_ht = [likelihood(1), 0.5*(likelihood(ind1) + likelihood(ind2)), likelihood(end)];
-                
+                likelihood = (diag((1 / sqrt((2 * pi)^ntemp*R(i, i)))* ...
+                    exp(-0.5*innovation.'*(R(i, i) \ innovation)))).';
+                likelihood = likelihood / min(likelihood); % for scaling
+                likelihoodht = [likelihood(1), 0.5 * (likelihood(ind1) + likelihood(ind2)), likelihood(end)];
```
Comment (AndreyAPopov) [inline src/+datools/+statistical/+ensemble/RHF.m:92]: why not just override likelihood instead of creating a new variable

File: src/+datools/+statistical/+ensemble/RHF.m [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+temp = dx*pdf;
+area = sum(temp(1:end-1));
+end
+
+function xp = riemannsearch(x, y, area)
+dx = abs(x(2) - x(1));
+darea = dx*y;
+cumarea = cumsum(darea);
+xp = x(find(cumarea - area>0, 1));
+end
```
[the remark this answers is not in the document]
Comment (AndreyAPopov) [reply src/+datools/+statistical/+ensemble/RHF.m:167]: There is still no newline here....

File: src/+datools/+statistical/+ensemble/Hybrid.m (hunk 7 above)
Comment (Abhat-93) [reply src/+datools/+statistical/+ensemble/Hybrid.m:40]: It was giving an error when i used Xf saying that xf is not defined. Let me take another look at it

File: src/+datools/+examples/+sandu/l63_experiments.m (hunk 1 above)
Comment (Abhat-93) [reply src/+datools/+examples/+sandu/l63_experiments.m:27]: just was making some changes in the plots so wanted something quick. I will revert it to 500 before committing.

File: src/+datools/+utils/+stat/KLDiv.m [hunk 13]
```diff
@@ -0,0 +1,43 @@
+%function to find the KL divergence of two given distribution
+% For more information look up: 
+% https://en.wikipedia.org/wiki/Kullback-Leibler_divergence
+
+%pdf1 = sample pdf
+%pdf2 = reference pdf(Uniform)
```
[the remark this answers is not in the document]
Comment (Abhat-93) [reply src/+datools/+utils/+stat/KLDiv.m:6]: I am not sure what to change here exactly

File: src/+datools/+examples/+sandu/l63_experiments.m (hunk 3 above)
Comment (Abhat-93) [reply src/+datools/+examples/+sandu/l63_experiments.m:82]: I have no idea. it was there in the original script from where i copied. I will remove it.

File: src/+datools/+examples/+sandu/l63_experiments.m (hunk 4 above)
Comment (Abhat-93) [reply src/+datools/+examples/+sandu/l63_experiments.m:246]: It was not plotting the heatmap

File: src/+datools/+statistical/+ensemble/Hybrid.m (hunk 7 above)
Comment (Abhat-93) [reply src/+datools/+statistical/+ensemble/Hybrid.m:40]: xf has been used in all the filter classes to denote the total forecast ensemble. Do you want me to change just for this filter or for every other filters too?

File: src/+datools/+statistical/+ensemble/RHF.m (hunk 11 above)
Comment (Abhat-93) [reply src/+datools/+statistical/+ensemble/RHF.m:92]: Done

File: src/+datools/+statistical/+ensemble/RHF.m (hunk 10 above)
Comment (Abhat-93) [reply src/+datools/+statistical/+ensemble/RHF.m:87]: Done

File: src/+datools/+statistical/+ensemble/RHF.m (hunk 9 above)
Comment (Abhat-93) [reply src/+datools/+statistical/+ensemble/RHF.m:82]: Done

File: src/+datools/+statistical/+ensemble/RHF.m (hunk 8 above)
Comment (Abhat-93) [reply src/+datools/+statistical/+ensemble/RHF.m:35]: Deleted

File: src/+datools/+statistical/+ensemble/RHF.m [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+                    prior_ht{1} = ones(1,discrete_pts)*1/((ensN+1) * gtl);
+                    prior_ht{length(xfs) + 1} = ones(1,discrete_pts)*1/((ensN+1) * gtl);
+                else
+                    fprintf('Error! Gaussian tail not provided!\n');
+                    return;
+                end
+                                
+                inn = y(i) - sort(Hxf(i,:));
+                ntemp = length(y(i));
+                likelihood = (diag((1/sqrt((2*pi)^ntemp*det(R(i,i)))) * ...
```
[the remark this answers is not in the document]
Comment (Abhat-93) [reply src/+datools/+statistical/+ensemble/RHF.m]: Done

File: src/+datools/+utils/+stat/RH.m [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    if nargin < 2
+        fprintf('need to pass the truth at this time step');
+        exit;
+    end
+    var = length(model.RankHistogram);
+        
+    xf = model.Ensemble;
+    [~,ensN] = size(xf);
+    
+    %rankVal = zeros(length(var), ensN+2);
```
[the remark this answers is not in the document]
Comment (Abhat-93) [reply src/+datools/+utils/+stat/RH.m]: Done

File: src/+datools/+statistical/+ensemble/RHF.m (hunk 12 above)
[the remark this answers is not in the document]
Comment (Abhat-93) [reply src/+datools/+statistical/+ensemble/RHF.m:167]: Done

File: src/+datools/+utils/+stat/KLDiv.m [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+%a = p(1);
+
+%KLVal =  sum(pdf1.*(log(pdf1./pdf2))) * sign(p(1));
+
+xs = linspace(0, 1, numel(pdf1));
+ys = numel(pdf1)*pdf1;
+
+J = @(a) sum( ((a * xs.^2 - a * xs + 0.25*a + 1) - ys).^2 );
+options = optimoptions(@fminunc, 'Display', 'none');
```
[the remark this answers is not in the document]
Comment (Abhat-93) [reply src/+datools/+utils/+stat/KLDiv.m:37]: Why?

File: src/+datools/+utils/+stat/KLDiv.m [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+    pdf2 = pdf2/sum(pdf2);
+end
+
+p = polyfit(1:1:length(pdf1), pdf1, 2);
+
+%p = polyfit(linspace(0, 1, numel(pdf1)), pdf1, 2);
+
+%a = p(1);
+
+%KLVal =  sum(pdf1.*(log(pdf1./pdf2))) * sign(p(1));
```
[the remark this answers is not in the document]
Comment (Abhat-93) [reply src/+datools/+utils/+stat/KLDiv.m:31]: Done

File: src/+datools/+statistical/+ensemble/RHF.m [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+                for kk = 2:ensN
+                    area(kk) = post_ht{kk} * abs(xfs(kk) - xfs(kk-1));
+                end
+                for ll = 1:ensN+1
+                    post_ht{ll} = post_ht{ll}/sum(area);
+                end
+                area = area/sum(area);
+                % find the updated posterior points/particles
+                tempa = findpos(post_ht, ensN, area, xfs, xleft, xright);
+                %[val, in2] = sort(sortin);
```
[the remark this answers is not in the document]
Comment (Abhat-93) [reply src/+datools/+statistical/+ensemble/RHF.m:117]: Done

File: src/+datools/+examples/+sandu/l63_experiments.m (hunk 2 above)
Comment (Abhat-93) [reply src/+datools/+examples/+sandu/l63_experiments.m:51]: Done

File: src/+datools/+examples/+sandu/l63_experiments.m (hunk 5 above)
Comment (Abhat-93) [reply src/+datools/+examples/+sandu/l63_experiments.m:251]: Done

File: src/+datools/+statistical/+ensemble/Hybrid.m (hunk 7 above)
Comment (AndreyAPopov) [reply src/+datools/+statistical/+ensemble/Hybrid.m:40]: I think amit did that... zzz. Whatever it's fine

**Comment by AndreyAPopov** [comment]
I guess this is ready. Do you think it is ready?

**Comment by Abhat-93** [comment]
I think this is ready except for the hybrid filter (which I am trying to understand) and the minor changes in plot which I have no idea how to do

**Comment by Abhat-93** [comment]
At least he can run some experiments with this.

**Comment by AndreyAPopov** [comment]
Hybrid filter has nothing to do with this branch though...

The plotting changes are also something you can figure out later.

I am merging.

This is main now.
